## Assignment 2.2 - Naive Bayes
    Submitted by: 
    Amandeep Kaur (21044104)                                                                                   
    Bhupesh Dod (21046099)

## Import the Libraries

In [1]:
import numpy as nm
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import scipy.stats as stats

import warnings
warnings.filterwarnings("ignore")


from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, f1_score
from sklearn.decomposition import PCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.naive_bayes import MultinomialNB as MNB
from sklearn.naive_bayes import ComplementNB as CNB

Import all the necessary Libraries.

## Import the Data

In [2]:
abalone_data = pd.read_csv("/Users/amandeepkaur/Desktop/ECE657A/Assignments /Assignment 2/ece657a-W23-asg2-datasets-jupyterdemo/abalone.csv",
names= ['Sex', 'Length', 'Diameter', 'Height', 'Whole_weight', 'Shucked_weight', 
'Viscera_weight','Shell_weight', 'Rings'])


wine_data1 = pd.read_csv("/Users/amandeepkaur/Desktop/ECE657A/Assignments /Assignment 2/ece657a-W23-asg2-datasets-jupyterdemo/winequality-red.csv",
 sep = ";")
wine_data1["colour"]=1
wine_data1 = wine_data1[list(('fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'colour', 'quality'))]

wine_data2 = pd.read_csv("/Users/amandeepkaur/Desktop/ECE657A/Assignments /Assignment 2/ece657a-W23-asg2-datasets-jupyterdemo/winequality-white.csv",
sep = ";")
wine_data2["colour"]=0
wine_data2 = wine_data2[list(('fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'colour', 'quality'))]

wine_data = pd.concat([wine_data1, wine_data2])

wine_data = wine_data.reset_index()

X_abalone = abalone_data.iloc[:, 1:-1]
y_abalone = abalone_data.iloc[:, -1]

X_wine = wine_data.iloc[:, :-1]
y_wine = wine_data.iloc[:, -1]


Import all the datasets and combine the wine datasets into a single dataset.

## Abalone Data

### MultiNomial NB without 5 Cross Validation

#### Abalone Raw

In [34]:
X_a_train_raw, X_a_test_raw, y_a_train_raw, y_a_test_raw = train_test_split(X_abalone, y_abalone, random_state= 45)


sc = StandardScaler()
X_a_train_raw.iloc[:, :] = sc.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = sc.transform(X_a_test_raw.iloc[:, :])

minmax = MinMaxScaler()
X_a_train_raw.iloc[:, :] = minmax.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = minmax.transform(X_a_test_raw.iloc[:, :])

# On Abalone Raw
mnb = MNB()
X_a_train_raw = mnb.fit(X_a_train_raw , y_a_train_raw)
y_a_pred_raw = mnb.predict(X_a_test_raw)
print("Accuracy with MultiNomial Naive Bayes on Abalone Raw is : %.2f" %(accuracy_score(y_a_pred_raw, y_a_test_raw)*100)+"%")

Accuracy with MultiNomial Naive Bayes on Abalone Raw is : 18.18%


#### Abalone PCA

In [35]:
X_a_train_pca, X_a_test_pca, y_a_train_pca, y_a_test_pca = train_test_split(X_abalone, y_abalone, random_state= 45)


X_a_train_pca = sc.fit_transform(X_a_train_pca)
X_a_test_pca = sc.transform(X_a_test_pca)

pca = PCA(0.95)
X_a_train_pca = pca.fit_transform(X_a_train_pca)
X_a_test_pca = pca.transform(X_a_test_pca)


X_a_train_pca = minmax.fit_transform(X_a_train_pca)
X_a_test_pca = minmax.transform(X_a_test_pca)

# On Abalone PCA
X_a_train_pca = mnb.fit(X_a_train_pca, y_a_train_pca)
y_a_pred_pca = mnb.predict(X_a_test_pca)
print("Accuracy with MultiNomial Naive Bayes on Abalone PCA is : %.2f" %(accuracy_score(y_a_pred_pca, y_a_test_pca)*100)+"%")

Accuracy with MultiNomial Naive Bayes on Abalone PCA is : 17.99%


#### Abalone LDA

In [66]:
X_a_train_lda, X_a_test_lda, y_a_train_lda, y_a_test_lda = train_test_split(X_abalone, y_abalone, random_state= 45)


X_a_train_lda.iloc[:, :] = sc.fit_transform(X_a_train_lda.iloc[:, :])
X_a_test_lda.iloc[:, :] = sc.transform(X_a_test_lda.iloc[:, :])

lda = LDA(n_components=4)
X_a_train_lda = lda.fit_transform(X_a_train_lda, y_a_train_lda)
X_a_test_lda = lda.transform(X_a_test_lda)

X_a_train_lda = minmax.fit_transform(X_a_train_lda)
X_a_test_lda = minmax.transform(X_a_test_lda)

#On Abalone LDA

X_a_train_lda = mnb.fit(X_a_train_lda, y_a_train_lda)
y_a_pred_lda = mnb.predict(X_a_test_lda)
print("Accuracy with MultiNomial Naive Bayes on Abalone LDA is : %.2f" %(accuracy_score(y_a_pred_lda, y_a_test_lda)*100)+"%")

Accuracy with MultiNomial Naive Bayes on Abalone LDA is : 17.99%


##### It is to be noted that the Accuracy for both PCA and LDA is the same for Abalone for MultiNomial Naive Bayes Classifier.

## Multinomial Using 5 fold with NB on Abalone

#### Abalone Raw

In [6]:
X_a_train_raw, X_a_test_raw, y_a_train_raw, y_a_test_raw = train_test_split(X_abalone, y_abalone, random_state= 45)

sc = StandardScaler()
X_a_train_raw.iloc[:, :] = sc.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = sc.transform(X_a_test_raw.iloc[:, :])

minmax = MinMaxScaler()
X_a_train_raw.iloc[:, :] = minmax.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = minmax.transform(X_a_test_raw.iloc[:, :])


# On Abalone Raw
accuracy_raw=cross_val_score(MNB(), X_a_train_raw , y_a_train_raw, cv=5, scoring='accuracy')
print("The mean accuracy with MultiNomial Naive Bayes on Abalone Raw using 5 fold is : %.2f" %(nm.mean(accuracy_raw)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Abalone Raw using 5 fold is : 16.19%


#### Abalone PCA

In [7]:
X_a_train_pca, X_a_test_pca, y_a_train_pca, y_a_test_pca = train_test_split(X_abalone, y_abalone, random_state= 45)

sc = StandardScaler()
X_a_train_pca = sc.fit_transform(X_a_train_pca)
X_a_test_pca = sc.transform(X_a_test_pca)

pca = PCA(0.95)
X_a_train_pca = pca.fit_transform(X_a_train_pca)
X_a_test_pca = pca.transform(X_a_test_pca)


X_a_train_pca = minmax.fit_transform(X_a_train_pca)
X_a_test_pca = minmax.transform(X_a_test_pca)

# On Abalone PCA
accuracy_pca=cross_val_score(MNB(), X_a_train_pca, y_a_train_pca, cv=5, scoring='accuracy')
print("The mean accuracy with MultiNomial Naive Bayes on Abalone PCA using 5 fold is : %.2f" %(nm.mean(accuracy_pca)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Abalone PCA using 5 fold is : 16.00%


#### Abalone LDA

In [65]:
X_a_train_lda, X_a_test_lda, y_a_train_lda, y_a_test_lda = train_test_split(X_abalone, y_abalone, random_state= 45)

sc = StandardScaler()
X_a_train_lda.iloc[:, :] = sc.fit_transform(X_a_train_lda.iloc[:, :])
X_a_test_lda.iloc[:, :] = sc.transform(X_a_test_lda.iloc[:, :])

lda = LDA(n_components=4)
X_a_train_lda = lda.fit_transform(X_a_train_lda, y_a_train_lda)
X_a_test_lda = lda.transform(X_a_test_lda)

X_a_train_lda = minmax.fit_transform(X_a_train_lda)
X_a_test_lda = minmax.transform(X_a_test_lda)


# On Abalone LCA
accuracy_lda=cross_val_score(MNB(), X_a_train_lda, y_a_train_lda, cv=5, scoring='accuracy')
nm.mean(accuracy_lda)
print("The mean accuracy with MultiNomial Naive Bayes on Abalone LDA using 5 fold is : %.2f" %(nm.mean(accuracy_lda)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Abalone LDA using 5 fold is : 16.00%


##### After applying 5 fold validation, the mean accuracy of PCA and LDA for MultiNomial Naive Bayes is still equal.

## Complement NB on Abalone without 5 Fold

#### Abalone Raw

In [30]:
X_a_train_raw, X_a_test_raw, y_a_train_raw, y_a_test_raw = train_test_split(X_abalone, y_abalone, random_state= 45)


X_a_train_raw.iloc[:, :] = sc.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = sc.transform(X_a_test_raw.iloc[:, :])


X_a_train_raw.iloc[:, :] = minmax.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = minmax.transform(X_a_test_raw.iloc[:, :])

# On Abalone Raw
cnb = CNB()
X_a_train_raw = cnb.fit(X_a_train_raw , y_a_train_raw)
y_a_pred_raw = cnb.predict(X_a_test_raw)
print("The mean accuracy with Complement Naive Bayes on Abalone Raw is : %.2f" %(accuracy_score(y_a_pred_raw, y_a_test_raw)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone Raw is : 19.43%


#### Abalone PCA

In [10]:
X_a_train_pca, X_a_test_pca, y_a_train_pca, y_a_test_pca = train_test_split(X_abalone, y_abalone, random_state= 45)

X_a_train_pca = sc.fit_transform(X_a_train_pca)
X_a_test_pca = sc.transform(X_a_test_pca)

pca = PCA(0.95)
X_a_train_pca = pca.fit_transform(X_a_train_pca)
X_a_test_pca = pca.transform(X_a_test_pca)


X_a_train_pca = minmax.fit_transform(X_a_train_pca)
X_a_test_pca = minmax.transform(X_a_test_pca)

# On Abalone PCA
X_a_train_pca = cnb.fit(X_a_train_pca, y_a_train_pca)
y_a_pred_pca = cnb.predict(X_a_test_pca)
print("The mean accuracy with Complement Naive Bayes on Abalone PCA is : %.2f" %(accuracy_score(y_a_pred_pca, y_a_test_pca)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone PCA is : 19.33%


#### Abalone LDA

In [64]:
X_a_train_lda, X_a_test_lda, y_a_train_lda, y_a_test_lda = train_test_split(X_abalone, y_abalone, random_state= 45)

X_a_train_lda.iloc[:, :] = sc.fit_transform(X_a_train_lda.iloc[:, :])
X_a_test_lda.iloc[:, :] = sc.transform(X_a_test_lda.iloc[:, :])

lda = LDA(n_components=4)
X_a_train_lda = lda.fit_transform(X_a_train_lda, y_a_train_lda)
X_a_test_lda = lda.transform(X_a_test_lda)

X_a_train_lda = minmax.fit_transform(X_a_train_lda)
X_a_test_lda = minmax.transform(X_a_test_lda)

#On Abalone LDA

X_a_train_lda = cnb.fit(X_a_train_lda, y_a_train_lda)
y_a_pred_lda = cnb.predict(X_a_test_lda)
print("The mean accuracy with Complement Naive Bayes on Abalone LDA is : %.2f" %(accuracy_score(y_a_pred_lda, y_a_test_lda)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone LDA is : 25.07%


##### For Complement Naive Bayes, the accuracy for Abalone decreases from Raw(19.43%) to PCA(19.33%), however, for LDA it is the maximum of 25.07%.  

## CNB with Using 5 Fold on Abalone

#### Abalone Raw

In [32]:
X_a_train_raw, X_a_test_raw, y_a_train_raw, y_a_test_raw = train_test_split(X_abalone, y_abalone, random_state= 45)


X_a_train_raw.iloc[:, :] = sc.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = sc.transform(X_a_test_raw.iloc[:, :])


X_a_train_raw.iloc[:, :] = minmax.fit_transform(X_a_train_raw.iloc[:, :])
X_a_test_raw.iloc[:, :] = minmax.transform(X_a_test_raw.iloc[:, :])

# On Abalone Raw
accuracy_raw=cross_val_score(CNB(), X_a_train_raw, y_a_train_raw, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Abalone Raw is : %.2f" %(nm.mean(accuracy_raw)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone Raw is : 17.56%


#### Abalone PCA

In [13]:
X_a_train_pca, X_a_test_pca, y_a_train_pca, y_a_test_pca = train_test_split(X_abalone, y_abalone, random_state= 45)

X_a_train_pca = sc.fit_transform(X_a_train_pca)
X_a_test_pca = sc.transform(X_a_test_pca)

pca = PCA(0.95)
X_a_train_pca = pca.fit_transform(X_a_train_pca)
X_a_test_pca = pca.transform(X_a_test_pca)


X_a_train_pca = minmax.fit_transform(X_a_train_pca)
X_a_test_pca = minmax.transform(X_a_test_pca)

# On Abalone PCA
accuracy_pca=cross_val_score(CNB(), X_a_train_pca, y_a_train_pca, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Abalone PCA is : %.2f" %(nm.mean(accuracy_pca)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone PCA is : 20.02%


#### Abalone LDA

In [63]:
X_a_train_lda, X_a_test_lda, y_a_train_lda, y_a_test_lda = train_test_split(X_abalone, y_abalone, random_state= 45)

X_a_train_lda.iloc[:, :] = sc.fit_transform(X_a_train_lda.iloc[:, :])
X_a_test_lda.iloc[:, :] = sc.transform(X_a_test_lda.iloc[:, :])

lda = LDA(n_components=4)
X_a_train_lda = lda.fit_transform(X_a_train_lda, y_a_train_lda)
X_a_test_lda = lda.transform(X_a_test_lda)

X_a_train_lda = minmax.fit_transform(X_a_train_lda)
X_a_test_lda = minmax.transform(X_a_test_lda)

#On Abalone LDA

accuracy_lda=cross_val_score(CNB(), X_a_train_lda, y_a_train_lda, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Abalone LDA is : %.2f" %(nm.mean(accuracy_lda)*100)+"%")

The mean accuracy with Complement Naive Bayes on Abalone LDA is : 23.53%


##### For Complement Naive Bayes using 5 fold, the accuracy increases from Raw(17.56%) to PCA(20.02%) to LDA(23.53%), however, using 5 fold the overall accuracy has dropped for all three datasets.

# Wine Data

### MultiNomial NB without 5 Fold

#### Wine Raw

In [47]:
X_w_train_raw, X_w_test_raw, y_w_train_raw, y_w_test_raw = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_raw.iloc[:, :-1] = sc.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = sc.transform(X_w_test_raw.iloc[:, :-1])


X_w_train_raw.iloc[:, :-1] = minmax.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = minmax.transform(X_w_test_raw.iloc[:, :-1])

# On Wine Raw
X_w_train_raw = mnb.fit(X_w_train_raw , y_w_train_raw)
y_w_pred_raw = mnb.predict(X_w_test_raw)
print("The mean accuracy with MultiNomial Naive Bayes on Wine Raw is : %.2f" %(accuracy_score(y_w_pred_raw, y_w_test_raw)*100)+"%")


The mean accuracy with MultiNomial Naive Bayes on Wine Raw is : 47.26%


#### Wine PCA

In [48]:
X_w_train_pca, X_w_test_pca, y_w_train_pca, y_w_test_pca = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_pca.iloc[:, :-1] = sc.fit_transform(X_w_train_pca.iloc[:, :-1])
X_w_test_pca.iloc[:, :-1] = sc.transform(X_w_test_pca.iloc[:, :-1])

pca = PCA(0.95)
X_w_train_pca = pca.fit_transform(X_w_train_pca)
X_w_test_pca = pca.transform(X_w_test_pca)


X_w_train_pca = minmax.fit_transform(X_w_train_pca)
X_w_test_pca = minmax.transform(X_w_test_pca)

# On Wine PCA
X_w_train_pca = mnb.fit(X_w_train_pca, y_w_train_pca)
y_w_pred_pca = mnb.predict(X_w_test_pca)
print("The mean accuracy with MultiNomial Naive Bayes on Wine PCA is : %.2f" %(accuracy_score(y_w_pred_pca, y_w_test_pca)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Wine PCA is : 41.23%


#### Wine LDA

In [49]:
X_w_train_lda, X_w_test_lda, y_w_train_lda, y_w_test_lda = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_lda.iloc[:, :-1] = sc.fit_transform(X_w_train_lda.iloc[:, :-1])
X_w_test_lda.iloc[:, :-1] = sc.transform(X_w_test_lda.iloc[:, :-1])

lda = LDA(n_components=4)
X_w_train_lda = lda.fit_transform(X_w_train_lda, y_w_train_lda)
X_w_test_lda = lda.transform(X_w_test_lda)

X_w_train_lda = minmax.fit_transform(X_w_train_lda)
X_w_test_lda = minmax.transform(X_w_test_lda)

#On Wine LDA

X_w_train_lda = mnb.fit(X_w_train_lda, y_w_train_lda)
y_w_pred_lda = mnb.predict(X_w_test_lda)
print("The mean accuracy with MultiNomial Naive Bayes on Wine lda is : %.2f" %(accuracy_score(y_w_pred_lda, y_w_test_lda)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Wine lda is : 41.23%


##### It is to be noted that the Accuracy for both PCA and LDA is the same for Wine for MultiNomial Naive Bayes Classifier.

## Complement NB

#### Wine RAW

In [50]:
X_w_train_raw, X_w_test_raw, y_w_train_raw, y_w_test_raw = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_raw.iloc[:, :-1] = sc.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = sc.transform(X_w_test_raw.iloc[:, :-1])


X_w_train_raw.iloc[:, :-1] = minmax.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = minmax.transform(X_w_test_raw.iloc[:, :-1])

# On Wine Raw

X_w_train_raw = cnb.fit(X_w_train_raw , y_w_train_raw)
y_w_pred_raw = cnb.predict(X_w_test_raw)
print("The accuracy with Complement Naive Bayes on Wine RAW is : %.2f" %(accuracy_score(y_w_pred_raw, y_w_test_raw)*100)+"%")


The accuracy with Complement Naive Bayes on Wine RAW is : 42.83%


#### Wine PCA

In [51]:
X_w_train_pca, X_w_test_pca, y_w_train_pca, y_w_test_pca = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_pca.iloc[:, :-1] = sc.fit_transform(X_w_train_pca.iloc[:, :-1])
X_w_test_pca.iloc[:, :-1] = sc.transform(X_w_test_pca.iloc[:, :-1])

pca = PCA(0.95)
X_w_train_pca = pca.fit_transform(X_w_train_pca)
X_w_test_pca = pca.transform(X_w_test_pca)


X_w_train_pca = minmax.fit_transform(X_w_train_pca)
X_w_test_pca = minmax.transform(X_w_test_pca)


# On Wine PCA

X_w_train_pca = cnb.fit(X_w_train_pca , y_w_train_pca)
y_w_pred_pca = cnb.predict(X_w_test_pca)
print("The accuracy with Complement Naive Bayes on Wine PCA is : %.2f" %(accuracy_score(y_w_pred_pca, y_w_test_pca)*100)+"%")

The accuracy with Complement Naive Bayes on Wine PCA is : 44.74%


#### Wine LDA

In [55]:
X_w_train_lda, X_w_test_lda, y_w_train_lda, y_w_test_lda = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_lda.iloc[:, :-1] = sc.fit_transform(X_w_train_lda.iloc[:, :-1])
X_w_test_lda.iloc[:, :-1] = sc.transform(X_w_test_lda.iloc[:, :-1])

lda = LDA(n_components=4)
X_w_train_lda = lda.fit_transform(X_w_train_lda, y_w_train_lda)
X_w_test_lda = lda.transform(X_w_test_lda)

X_w_train_lda = minmax.fit_transform(X_w_train_lda)
X_w_test_lda = minmax.transform(X_w_test_lda)


#On Wine LDA

X_w_train_lda = cnb.fit(X_w_train_lda, y_w_train_lda)
y_w_pred_lda = cnb.predict(X_w_test_lda)
print("The accuracy with Complement Naive Bayes on Wine LDA is : %.2f" %(accuracy_score(y_w_pred_lda, y_w_test_lda)*100)+"%")

The accuracy with Complement Naive Bayes on Wine LDA is : 44.68%


##### For Complement Naive Bayes, the accuracy increases from Raw(42.83%) to PCA(44.74%), to LDA(44.68%).

## Using 5 Fold Validation with MNB on Wine

#### Wine RAW

In [60]:
X_w_train_raw, X_w_test_raw, y_w_train_raw, y_w_test_raw = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_raw.iloc[:, :-1] = sc.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = sc.transform(X_w_test_raw.iloc[:, :-1])


X_w_train_raw.iloc[:, :-1] = minmax.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = minmax.transform(X_w_test_raw.iloc[:, :-1])

# On Wine Raw

accuracy_raw=cross_val_score(MNB(), X_w_train_raw, y_w_train_raw, cv=5, scoring='accuracy')
print("The mean accuracy with MultiNomial Naive Bayes on Wine RAW with 5 fold is : %.2f" %(nm.mean(accuracy_raw)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Wine RAW with 5 fold is : 47.89%


#### Wine PCA

In [61]:
X_w_train_pca, X_w_test_pca, y_w_train_pca, y_w_test_pca = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_pca.iloc[:, :-1] = sc.fit_transform(X_w_train_pca.iloc[:, :-1])
X_w_test_pca.iloc[:, :-1] = sc.transform(X_w_test_pca.iloc[:, :-1])

pca = PCA(0.95)
X_w_train_pca = pca.fit_transform(X_w_train_pca)
X_w_test_pca = pca.transform(X_w_test_pca)


X_w_train_pca = minmax.fit_transform(X_w_train_pca)
X_w_test_pca = minmax.transform(X_w_test_pca)

# On Wine PCA
accuracy_pca=cross_val_score(MNB(), X_w_train_pca, y_w_train_pca, cv=5, scoring='accuracy')

print("The mean accuracy with MultiNomial Naive Bayes on Wine PCA with 5 fold is : %.2f" %(nm.mean(accuracy_pca)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Wine PCA with 5 fold is : 44.46%


#### Wine LDA

In [62]:
X_w_train_lda, X_w_test_lda, y_w_train_lda, y_w_test_lda = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_lda.iloc[:, :-1] = sc.fit_transform(X_w_train_lda.iloc[:, :-1])
X_w_test_lda.iloc[:, :-1] = sc.transform(X_w_test_lda.iloc[:, :-1])

lda = LDA(n_components=4)
X_w_train_lda = lda.fit_transform(X_w_train_lda, y_w_train_lda)
X_w_test_lda = lda.transform(X_w_test_lda)

X_w_train_lda = minmax.fit_transform(X_w_train_lda)
X_w_test_lda = minmax.transform(X_w_test_lda)

#On Wine LDA

accuracy_lda=cross_val_score(MNB(), X_w_train_lda, y_w_train_lda, cv=5, scoring='accuracy')
print("The mean accuracy with MultiNomial Naive Bayes on Wine LDA with 5 fold is : %.2f" %(nm.mean(accuracy_lda)*100)+"%")

The mean accuracy with MultiNomial Naive Bayes on Wine LDA with 5 fold is : 44.46%


##### After applying 5 fold validation, the mean accuracy of PCA and LDA for MultiNomial Naive Bayes is still equal. However, it is less than the RAW Accuracy.

## Using 5 Fold Validation on CNB

#### Wine RAW

In [56]:
X_w_train_raw, X_w_test_raw, y_w_train_raw, y_w_test_raw = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_raw.iloc[:, :-1] = sc.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = sc.transform(X_w_test_raw.iloc[:, :-1])


X_w_train_raw.iloc[:, :-1] = minmax.fit_transform(X_w_train_raw.iloc[:, :-1])
X_w_test_raw.iloc[:, :-1] = minmax.transform(X_w_test_raw.iloc[:, :-1])

# On Wine Raw

accuracy_raw=cross_val_score(CNB(), X_w_train_raw, y_w_train_raw, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Wine RAW using 5 fold is : %.2f" %(nm.mean(accuracy_raw)*100)+"%")


The mean accuracy with Complement Naive Bayes on Wine RAW using 5 fold is : 44.44%


#### Wine PCA

In [57]:
X_w_train_pca, X_w_test_pca, y_w_train_pca, y_w_test_pca = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_pca.iloc[:, :-1] = sc.fit_transform(X_w_train_pca.iloc[:, :-1])
X_w_test_pca.iloc[:, :-1] = sc.transform(X_w_test_pca.iloc[:, :-1])

pca = PCA(0.95)
X_w_train_pca = pca.fit_transform(X_w_train_pca)
X_w_test_pca = pca.transform(X_w_test_pca)


X_w_train_pca = minmax.fit_transform(X_w_train_pca)
X_w_test_pca = minmax.transform(X_w_test_pca)

# On Wine PCA
accuracy_pca=cross_val_score(CNB(), X_w_train_pca, y_w_train_pca, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Wine PCA using 5 fold is : %.2f" %(nm.mean(accuracy_pca)*100)+"%")

The mean accuracy with Complement Naive Bayes on Wine PCA using 5 fold is : 43.08%


#### Wine PCA

In [58]:
X_w_train_lda, X_w_test_lda, y_w_train_lda, y_w_test_lda = train_test_split(X_wine, y_wine, random_state= 45)


X_w_train_lda.iloc[:, :-1] = sc.fit_transform(X_w_train_lda.iloc[:, :-1])
X_w_test_lda.iloc[:, :-1] = sc.transform(X_w_test_lda.iloc[:, :-1])

lda = LDA(n_components=4)
X_w_train_lda = lda.fit_transform(X_w_train_lda, y_w_train_lda)
X_w_test_lda = lda.transform(X_w_test_lda)

X_w_train_lda = minmax.fit_transform(X_w_train_lda)
X_w_test_lda = minmax.transform(X_w_test_lda)

#On Wine LDA

accuracy_lda=cross_val_score(CNB(), X_w_train_lda, y_w_train_lda, cv=5, scoring='accuracy')
print("The mean accuracy with Complement Naive Bayes on Wine LDA using 5 fold is : %.2f" %(nm.mean(accuracy_lda)*100)+"%")

The mean accuracy with Complement Naive Bayes on Wine LDA using 5 fold is : 42.59%


##### For Complement Naive Bayes using 5 fold, the accuracy decerease from Raw(44.44%) to PCA(43.08%) to LDA(42.59%), however, using 5 fold the overall accuracy has dropped for all three datasets.

In [27]:
abalone_accuracy_array = {'Abalone_Raw_MNB' : [16.19, 18.18],
                    'Abalone_PCA_MNB' : [16, 17.99],
                    'Abalone_LDA_MNB' : [16, 17.99],
                    'Abalone_Raw_CNB' : [17.56, 19.43],
                    'Abalone_PCA_CNB' : [20.02, 19.33],
                    'Abalone_LDA_CNB' : [23.53, 25.07]}

pd.DataFrame(abalone_accuracy_array, index = ['With 5 Fold', 'Without 5 Fold'])

Abalone_Raw_MNB  Abalone_PCA_MNB  Abalone_LDA_MNB  \
With 5 Fold               16.19            16.00            16.00   
Without 5 Fold            18.18            17.99            17.99   

                Abalone_Raw_CNB  Abalone_PCA_CNB  Abalone_LDA_CNB  
With 5 Fold               17.56            20.02            23.53  
Without 5 Fold            19.43            19.33            25.07

In [59]:
wine_accuracy_array = {'Wine_Raw_MNB' : [47.89, 47.26],
                    'Wine_PCA_MNB' : [44.46, 41.23],
                    'Wine_LDA_MNB' : [44.46, 41.23],
                    'Wine_Raw_CNB' : [44.4, 42.83],
                    'Wine_PCA_CNB' : [43.08,44.74],
                    'Wine_LDA_CNB' : [42.59, 44.68]}

pd.DataFrame(wine_accuracy_array, index = ['With 5 Fold', 'Without 5 Fold'])

Wine_Raw_MNB  Wine_PCA_MNB  Wine_LDA_MNB  Wine_Raw_CNB  \
With 5 Fold            47.89         44.46         44.46         44.40   
Without 5 Fold         47.26         41.23         41.23         42.83   

                Wine_PCA_CNB  Wine_LDA_CNB  
With 5 Fold            43.08         42.59  
Without 5 Fold         44.74         44.68

#### The Accuracies for MNB Classifier remained same for both PCA and LDA in both the datasets.